### ISSUE TITLE
False positive `[operator]` on union of dicts with heterogeneous value types

### ISSUE THREAD

**Issue by harahu** [python/mypy#18236]
**Bug Report**

If I take the union of two dicts with different value types, I cannot assign the result to a variable that is annotated with the resulting type. Doing so causes an error. Assigning to an non-annotated variable does not cause an error, which seems like the correct behavior.

Do I really need the annotation on `d5` below? No. But I have a non-toy example of this in a code base where `d5` is a type annotated attribute on an object, which results in the same error.

```python
d1: dict[str, str] = {}
d2: dict[str, str | None] = {}

d3 = d1 | d2
reveal_type(d3)  # builtins.dict[builtins.str, Union[builtins.str, None]], aka dict[str, str | None]

# This is fine
d4: dict[str, str | None] = d3  # aka d1 | d2

# This isn't?
d5: dict[str, str | None] = d1 | d2
```

**To Reproduce**

See: https://mypy-play.net/?mypy=master&python=3.13&flags=show-error-codes&gist=2643cb3ff1ea956a1c4ed0869140d46d

**Expected Behavior**

I expect `d5: dict[str, str | None] = d1 | d2` to be legal.

**Actual Behavior**

`error: Unsupported operand types for | ("dict[str, str]" and "dict[str, str | None]")  [operator]`.

**Your Environment**

- Mypy version used: 1.13.0
- Mypy command-line flags:
- Mypy configuration options from `mypy.ini` (and other config files):
- Python version used: 3.12

**Comment by Aarya2004** [comment]
Hey @harahu, I'd love to work on this! Thanks for including the code snippet, I'm looking through the code for ways to make a fix. Let me know if you have any possible solutions, I'll let you know if I have any updates!

**Comment by brianschubert** [comment]
@Aarya2004 Thanks for your interest in contributing to mypy! Unfortunately, this probably isn't a good issue for a first contribution. This issue has to do with they way mypy applies bidirectional inference, which  a) is a very complex part of the code, and b) involves making tradeoffs that are tricky to get right and have a high risk of breaking behavior that mypy users rely on.

I'd recommend taking a look at issue tagged https://github.com/python/mypy/labels/good-first-issue or https://github.com/python/mypy/labels/good-second-issue, which should be a bit more approachable and are a good way to build up to larger issues. You can also check out [CONTRIBUTING.md](https://github.com/python/mypy/blob/master/CONTRIBUTING.md) and the [Developer Guides](https://github.com/python/mypy/wiki/Developer-Guides) for contributing guidelines and tips on where to start.

**Comment by Aarya2004** [comment]
Thanks for the guidance @brianschubert. I'll take a look at that then!
